Firestore/core/src/firebase/firestore/remote/remote_event.mm

```diff
@@ -0,0 +1,417 @@
+  //
+  // TODO(gsoltis): Expand on this comment.
+  for (const auto& entry : pending_document_target_mappings_) {
+    bool is_only_limbo_target = true;
+
+    for (TargetId target_id : entry.second) {
+      FSTQueryData* query_data = QueryDataForActiveTarget(target_id);
+      if (query_data && query_data.purpose != FSTQueryPurposeLimboResolution) {
+        is_only_limbo_target = false;
+        break;
+      }
+    }
+
+    if (is_only_limbo_target) {
+      resolved_limbo_documents = resolved_limbo_documents.insert(entry.first);
+    }
+  }
+
+  FSTRemoteEvent* remote_event =
+      [[FSTRemoteEvent alloc] initWithSnapshotVersion:snapshot_version
+                                        targetChanges:target_changes
+                                     targetMismatches:pending_target_resets_
+                                      documentUpdates:pending_document_updates_
+                                       limboDocuments:resolved_limbo_documents];
+
+  // Re-initialize the current state to ensure that we do not modify the
+  // generated `RemoteEvent`.
+  pending_document_updates_.clear();
+  pending_document_target_mappings_.clear();
+  pending_target_resets_.clear();
+
+  return remote_event;
+}
+
+void WatchChangeAggregator::AddDocumentToTarget(TargetId target_id,
+                                                FSTMaybeDocument* document) {
+  if (!IsActiveTarget(target_id)) {
+    return;
+  }
+
+  DocumentViewChangeType change_type =
+      TargetContainsDocument(target_id, document.key)
+          ? DocumentViewChangeType::kModified
+          : DocumentViewChangeType::kAdded;
+
+  TargetState& target_state = EnsureTargetState(target_id);
+  target_state.AddDocumentChange(document.key, change_type);
+
+  pending_document_updates_[document.key] = document;
+  pending_document_target_mappings_[document.key].insert(target_id);
+}
+
+void WatchChangeAggregator::RemoveDocumentFromTarget(
+    TargetId target_id,
+    const DocumentKey& key,
+    FSTMaybeDocument* _Nullable updated_document) {
+  if (!IsActiveTarget(target_id)) {
+    return;
+  }
+
+  TargetState& target_state = EnsureTargetState(target_id);
+  if (TargetContainsDocument(target_id, key)) {
+    target_state.AddDocumentChange(key, DocumentViewChangeType::kRemoved);
+  } else {
+    // The document may have entered and left the target before we raised a
+    // snapshot, so we can just ignore the change.
+    target_state.RemoveDocumentChange(key);
+  }
+  pending_document_target_mappings_[key].insert(target_id);
+
+  if (updated_document) {
+    pending_document_updates_[key] = updated_document;
+  }
+}
+
+void WatchChangeAggregator::RemoveTarget(TargetId target_id) {
+  target_states_.erase(target_id);
+}
+
+int WatchChangeAggregator::GetCurrentDocumentCountForTarget(
+    TargetId target_id) {
+  TargetState& target_state = EnsureTargetState(target_id);
+  FSTTargetChange* target_change = target_state.ToTargetChange();
```

Comment:
This is pretty inefficient (creating a target change involves creating and filling three tree-based immutable maps), though it's only called from `HandleExistenceFilter` which probably doesn't happen all that often.